charms/istio-pilot/src/charm.py

```diff
@@ -136,6 +145,64 @@ def install(self, event):
 
         self.unit.status = ActiveStatus()
 
+    def reconcile(self, event):
+        """Reconcile the state of the charm.
+
+        TODO: Add more details
+        """
+        # If we are not the leader, the charm should do nothing else
+        try:
+            self._check_leader()
+        except ErrorWithStatus as err:
+            self._log_and_set_status(err.status)
+            return
+
+        # This charm may hit multiple, non-fatal errors during the reconciliation.  Collect them
+        # so that we can report them at the end.
+        handled_errors = []
+
+        ingress_auth_reconcile_successful = False
+        try:
+            ingress_auth = self._get_ingress_auth_data()
+            self._reconcile_ingress_auth(ingress_auth)
+            ingress_auth_reconcile_successful = True
+        except ErrorWithStatus as err:
+            handled_errors.append(err)
+
+        try:
+            # If handling the ingress_auth relation fails, remove the Gateway to prevent
+            # unauthenticated traffic
+            if ingress_auth_reconcile_successful:
+                self._reconcile_gateway()
+            else:
+                # TODO: Log here?
+                self._remove_gateway()
+        except ErrorWithStatus as err:
+            handled_errors.append(err)
+
+        try:
+            self._send_gateway_info()
+        except ErrorWithStatus as err:
+            handled_errors.append(err)
+
+        try:
+            # TODO: Should I break these up so we can have more granular behaviour?
+            #  I could just get the ingress interface directly.  Although I can't break that
+            #  up into its components easily due to how SDI is written.
+            # If any relation in this group has a version error, this will fail fast and not
+            # provide any data for us to work on.  This is a limitation of SDI.
+            interfaces = self._get_interfaces()
+            self._reconcile_ingress(interfaces['ingress'], event)
+        except ErrorWithStatus as err:
+            # One or more related applications resulted in an error
+            handled_errors.append(err)
+            self._log_and_set_status(err.status)
+
+        # TODO: If we have and handled_errors, report them
+        self._report_handled_errors(handled_errors)
+
```

Comment:
Where are we going to raise the "worst" error?

Reply:
I'm thinking this handler does something like:
```python
def _report_handled_errors(self, handled_errors):
  worst_error = get_worst_error(handled_errors)
  worst_error_type = worst_error.class  # Or however you get this...
  headline_message = f"During operation we encountered {len(handled_errors)} errors.  See logs for details."
  self.unit.status = worst_error_type(f"During operation we encountered {len(handled_errors)} handled errors.  See logs for details.")
  self.log.info(f"During operation we encountered {len(handled_errors)} handled errors:")

  for error in handled_errors:
    self.log.info(f"Handled error of type {error.status}: {error.message}")
```

what do you think?

Reply:
Look good. I think, we should use `log.error` though. Also, do we need to explicitly call `raise` for that "worst" error?